proteuslib/property_models/tests/property_test_harness.py

```diff
@@ -0,0 +1,447 @@
+                                              number_total_constraints,
+                                              number_unused_variables)
+from idaes.core.util.scaling import (calculate_scaling_factors,
+                                     unscaled_variables_generator,
+                                     unscaled_constraints_generator,
+                                     badly_scaled_var_generator)
+from idaes.core.util.initialization import fix_state_vars
+from idaes.core.util import get_solver
+
+
+# -----------------------------------------------------------------------------
+class PropertyTestHarness(object):
+    def configure_class(self, m):
+        self.configure()
+
+        # attaching objects to model to carry through in pytest frame
+        assert not hasattr(m, '_test_objs')
+        m._test_objs = Block()
+        m._test_objs.stateblock_statistics = self.stateblock_statistics
+        m._test_objs.default_solution = self.default_solution
+
+    def configure(self):
+        """
+        Placeholder method to allow user to setup test harness.
+
+        The configure function must set the attributes:
+
+        prop_pack: property package parameter block
+
+        param_args: dictionary for property parameter arguments
+
+        scaling_args: dictionary for standard scaling arguments
+            keys = (string name of variable, tuple index), values = scaling factor
+
+        stateblock_statistics: dictionary of model statistics
+            {'number_variables': VALUE,
+             'number_total_constraints': VALUE,
+             'number_unused_variables': VALUE,
+             'default_degrees_of_freedom': VALUE}
+
+        default_solution: dictionary of property values at default initialized state variables
+            keys = (string name of variable, tuple index), values = value
+        """
+        pass
+
+    @pytest.fixture(scope='class')
+    def frame_stateblock(self):
+        m = ConcreteModel()
+        self.configure_class(m)
+
+        m.fs = FlowsheetBlock(default={"dynamic": False})
+        m.fs.properties = self.prop_pack()
+        m.fs.stream = m.fs.properties.build_state_block([0], default=self.param_args)
+
+        for (v_str, ind), sf in self.scaling_args.items():
+            m.fs.properties.set_default_scaling(v_str, sf, index=ind)
+
+        return m
+
+    @pytest.mark.unit
+    def test_components_phases(self, frame_stateblock):
+        m = frame_stateblock
+        assert isinstance(m.fs.properties.component_list, Set)
+        assert isinstance(m.fs.properties.phase_list, Set)
+        assert isinstance(m.fs.properties._phase_component_set, Set)
+
+    @pytest.mark.unit
+    def test_parameters(self, frame_stateblock):
+        m = frame_stateblock
+        # test that the parameter variables are fixed
+        for v in m.fs.properties.component_data_objects(Var):
+            if not v.is_fixed():
+                raise Exception(
```

Comment:
You should create your own subclass of Exception, e.g. PropertyTestException. You can create multiple subclasses derived from that if you want to get even more specific.